File: apps/quickink/android/app/src/main/java/app/quickink/mobile/features/notes/NoteEditorScreen.kt
/*
 * NoteEditorScreen.kt
 *
 * QuickInk's Note Detail — upgraded from the slim title+body editor
 * to the mockup's tabbed view:
 *
 *   - Tab 1 (Page): lined-paper page rendering with coral left
 *     margin; handwritten Caveat preview of the note body. When
 *     a real captured image is wired through, the image overlays
 *     the paper rule.
 *
 *   - Tab 2 (OCR Text): clean editorial serif rendering of the
 *     transcript with a confidence badge, copy-to-clipboard button,
 *     and a "smart suggestions" footer hook (currently inert).
 *
 *   - Floating action bar (Re-tag / Export / Delete).
 *
 * Architecturally still wraps `NoteEditorController` — controller
 * fields `title` and `notes` remain the data source.
 *
 * Mirror of iOS `NoteEditorScreen.swift`.
 */

package app.quickink.mobile.features.notes

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.slideInVertically
import androidx.compose.animation.slideOutVertically
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.ContentCopy
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material.icons.filled.IosShare
import androidx.compose.material.icons.filled.LocalOffer
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.quickink.mobile.QuickInkApp
import app.quickink.mobile.data.sync.QuickInkSyncScheduler
import app.quickink.mobile.features.settings.SettingsPreferences
import app.quickink.mobile.ui.theme.LocalQuickInkColors
import app.quickink.mobile.ui.theme.LocalQuickInkTypography
import app.quickink.mobile.ui.theme.QuickInkRadius
import app.quickink.mobile.ui.theme.QuickInkSpacing
import app.quickink.mobile.ui.theme.quickInkDotGridBackground
import app.quickink.mobile.ui.theme.quickInkLinedPaper
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

private enum class DetailTab { Page, OCR }

@Composable
fun NoteEditorScreen(
    entryId: String,
    userId: String,
    onBack: () -> Unit,
) {
    val context = LocalContext.current
    val app = context.applicationContext as QuickInkApp
    val scope = rememberCoroutineScope()
    val colors = LocalQuickInkColors.current
    val type = LocalQuickInkTypography.current

    val controller = remember(entryId) {
        NoteEditorController(
            entryId   = entryId,
            userId    = userId,
            dao       = app.database.notepadDao(),
            scope     = scope,
            // Sync is user-initiated only — no auto-kick on note
            // edits. User taps Settings → "Sync now" to push.
            onMutated = { /* intentional no-op */ },
        )
    }

    LaunchedEffect(controller) {
        controller.bootstrap()
    }

    var activeTab by remember { mutableStateOf(DetailTab.Page) }
    var copyToastVisible by remember { mutableStateOf(false) }
    var showExportSheet by remember { mutableStateOf(false) }
    val searchablePdfEnabled = remember {
        SettingsPreferences(context).searchablePdfExportEnabled
    }

    Box(modifier = Modifier.fillMaxSize().quickInkDotGridBackground()) {
        Column(modifier = Modifier.fillMaxSize()) {
            DetailTopBar(
                title    = controller.title,
                date     = controller.entryDate,
                onBack   = {
                    if (controller.canSave) controller.save()
                    onBack()
                },
            )

            TabSwitcher(
                active   = activeTab,
                onSelect = { activeTab = it },
                modifier = Modifier
                    .padding(horizontal = QuickInkSpacing.s5)
                    .padding(top = QuickInkSpacing.s3, bottom = QuickInkSpacing.s2),
            )

            if (controller.isLoading) {
                Box(
                    modifier = Modifier.fillMaxSize(),
                    contentAlignment = Alignment.Center,
                ) {
                    CircularProgressIndicator(color = colors.accent)
                }
            } else {
                when (activeTab) {
                    DetailTab.Page -> PageTab(controller = controller)
                    DetailTab.OCR  -> OcrTab(
                        controller = controller,
                        onCopy     = {
                            val cm = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
                            cm.setPrimaryClip(ClipData.newPlainText("OCR", controller.notes))
                            copyToastVisible = true
                            scope.launch {
                                delay(1600)
                                copyToastVisible = false
                            }
                        },
                    )
                }
            }
        }

        FloatingActionBar(
            showDelete = controller.entry != null,
            onReTag    = { /* T13 */ },
            onExport   = { showExportSheet = true },
            onDelete   = { controller.delete(onDeleted = onBack) },
            modifier   = Modifier
                .align(Alignment.BottomCenter)
                .padding(horizontal = QuickInkSpacing.s5)
                .padding(bottom = QuickInkSpacing.s5),
        )

        if (showExportSheet) {
            ExportSheet(
                searchablePdfEnabled = searchablePdfEnabled,
                onSelect = { format ->
                    showExportSheet = false
                    // Format-specific export pipeline lands in a
                    // follow-up. Currently a no-op — the picker
                    // surfaces intent without generating the file.
                    @Suppress("UNUSED_EXPRESSION") format
                },
                onDismiss = { showExportSheet = false },
            )
        }

        AnimatedVisibility(
            visible = copyToastVisible,
            enter   = fadeIn() + slideInVertically(initialOffsetY  = { it / 2 }),
            exit    = fadeOut() + slideOutVertically(targetOffsetY = { it / 2 }),
            modifier = Modifier
                .align(Alignment.BottomCenter)
                .padding(bottom = 100.dp),
        ) {
            Box(
                modifier = Modifier
                    .clip(RoundedCornerShape(QuickInkRadius.pill))
                    .background(colors.ink)
                    .padding(horizontal = QuickInkSpacing.s4, vertical = QuickInkSpacing.s2),
            ) {
                Text(
                    text  = "Copied to clipboard",
                    style = type.label,
                    color = colors.textOnAccent,
                )
            }
        }
    }
}

// MARK: - Top bar

@Composable
private fun DetailTopBar(
    title: String,
    date: String,
    onBack: () -> Unit,
) {
    val colors = LocalQuickInkColors.current
    val type = LocalQuickInkTypography.current
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = QuickInkSpacing.s2, vertical = QuickInkSpacing.s2),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        IconButton(onClick = onBack) {
            Icon(
                imageVector       = Icons.AutoMirrored.Filled.ArrowBack,
                contentDescription = "Back",
                tint              = colors.ink,
            )
        }
        Column(modifier = Modifier.weight(1f)) {
            Text(
                text     = title.ifEmpty { "Untitled" },
                style    = type.label,
                color    = if (title.isEmpty()) colors.muted else colors.ink,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
            Text(text = date, style = type.caption, color = colors.muted)
        }
        // Page navigator pill — placeholder showing 1/1.
        Box(
            modifier = Modifier
                .clip(RoundedCornerShape(QuickInkRadius.pill))
                .background(colors.borderSoft)
                .padding(horizontal = QuickInkSpacing.s3, vertical = 4.dp),
        ) {
            Text(text = "1 / 1", style = type.caption, color = colors.inkSoft)
        }
        Spacer(Modifier.size(QuickInkSpacing.s3))
    }
}

// MARK: - Tab switcher

@Composable
private fun TabSwitcher(
    active: DetailTab,
    onSelect: (DetailTab) -> Unit,
    modifier: Modifier = Modifier,
) {
    val colors = LocalQuickInkColors.current

    Row(
        modifier = modifier
            .clip(RoundedCornerShape(QuickInkRadius.pill))
            .background(colors.borderSoft)
            .padding(4.dp),
    ) {
        TabPill(label = "Page",     isActive = active == DetailTab.Page) { onSelect(DetailTab.Page) }
        Spacer(Modifier.size(4.dp))
        TabPill(label = "OCR Text", isActive = active == DetailTab.OCR)  { onSelect(DetailTab.OCR) }
    }
}

@Composable
private fun TabPill(label: String, isActive: Boolean, onClick: () -> Unit) {
    val colors = LocalQuickInkColors.current
    val type = LocalQuickInkTypography.current
    Box(
        modifier = Modifier
            .clip(RoundedCornerShape(QuickInkRadius.pill))
            .background(if (isActive) colors.surface else Color.Transparent)
            .clickable(onClick = onClick)
            .padding(horizontal = QuickInkSpacing.s4, vertical = QuickInkSpacing.s2),
    ) {
        Text(
            text  = label,
            style = type.label,
            color = if (isActive) colors.ink else colors.inkSoft,
        )
    }
}

// MARK: - Page tab

@Composable
private fun PageTab(controller: NoteEditorController) {
    val colors = LocalQuickInkColors.current
    val type = LocalQuickInkTypography.current

    Column(
        modifier = Modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState())
            .padding(QuickInkSpacing.s5)
            .padding(bottom = 100.dp),
    ) {
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .heightIn(min = 480.dp)
                .shadow(8.dp, RoundedCornerShape(QuickInkRadius.lg))
                .clip(RoundedCornerShape(QuickInkRadius.lg))
                .background(colors.surface)
                .border(1.dp, colors.border, RoundedCornerShape(QuickInkRadius.lg)),
        ) {
            // Lined paper rule lines.
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .quickInkLinedPaper(seed = controller.entryDate.hashCode(), lineSpacing = 22.dp, lineOpacity = 0.10f),
            )
            // Coral left margin line.
            Box(
                modifier = Modifier
                    .padding(start = 36.dp, top = 12.dp, bottom = 12.dp)
                    .width(1.5.dp)
                    .fillMaxSize()
                    .background(colors.accent.copy(alpha = 0.6f)),
            )
            // Content rendered in handwritten Caveat.
            Column(
                modifier = Modifier
                    .padding(start = 52.dp, end = QuickInkSpacing.s5, top = QuickInkSpacing.s5, bottom = QuickInkSpacing.s5),
                verticalArrangement = Arrangement.spacedBy(QuickInkSpacing.s3),
            ) {
                if (controller.title.isNotEmpty()) {
                    Text(
                        text  = controller.title,
                        style = type.handwritten.copy(fontSize = 28.sp),
                        color = colors.ink,
                    )
                }
                Text(
                    text = if (controller.notes.isEmpty()) "Tap OCR Text to add transcript content." else controller.notes,
                    style = type.handwritten.copy(fontSize = 20.sp),
                    color = colors.ink.copy(alpha = 0.85f),
                )
            }
        }
    }
}

// MARK: - OCR tab

@Composable
private fun OcrTab(controller: NoteEditorController, onCopy: () -> Unit) {
    val colors = LocalQuickInkColors.current
    val type = LocalQuickInkTypography.current

    Column(
        modifier = Modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState())
            .padding(horizontal = QuickInkSpacing.s5)
            .padding(top = QuickInkSpacing.s3, bottom = 100.dp),
        verticalArrangement = Arrangement.spacedBy(QuickInkSpacing.s4),
    ) {
        // Confidence + copy header.
        Row(verticalAlignment = Alignment.CenterVertically) {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier
                    .clip(RoundedCornerShape(QuickInkRadius.pill))
                    .background(colors.borderSoft)
                    .padding(horizontal = QuickInkSpacing.s3, vertical = QuickInkSpacing.s1),
                horizontalArrangement = Arrangement.spacedBy(QuickInkSpacing.s1),
            ) {
                Box(
                    modifier = Modifier.size(6.dp).clip(CircleShape).background(colors.success),
                )
                Text(text = "98% confidence", style = type.caption, color = colors.inkSoft)
            }

            Spacer(Modifier.weight(1f))

            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier
                    .clip(RoundedCornerShape(QuickInkRadius.pill))
                    .background(colors.accentSoft)
                    .clickable(onClick = onCopy)
                    .padding(horizontal = QuickInkSpacing.s3, vertical = QuickInkSpacing.s1),
                horizontalArrangement = Arrangement.spacedBy(6.dp),
            ) {
                Icon(
                    imageVector       = Icons.Filled.ContentCopy,
                    contentDescription = null,
                    tint              = colors.accent,
                    modifier          = Modifier.size(13.dp),
                )
                Text(text = "Copy", style = type.label, color = colors.accent)
            }
        }

        // Title.
        BasicTextField(
            value         = controller.title,
            onValueChange = { controller.title = it },
            textStyle     = type.pageTitle.copy(color = colors.ink),
            cursorBrush   = SolidColor(colors.accent),
            singleLine    = true,
            modifier      = Modifier.fillMaxWidth(),
            decorationBox = { inner ->
                if (controller.title.isEmpty()) {
                    Text(text = "Title", style = type.pageTitle, color = colors.muted)
                }
                inner()
            },
        )

        // Body / OCR transcript.
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(QuickInkRadius.md))
                .background(colors.surface)
                .border(1.dp, colors.border, RoundedCornerShape(QuickInkRadius.md))
                .padding(QuickInkSpacing.s4),
        ) {
            BasicTextField(
                value         = controller.notes,
                onValueChange = { controller.notes = it },
                textStyle     = type.body.copy(color = colors.ink),
                cursorBrush   = SolidColor(colors.accent),
                modifier      = Modifier
                    .fillMaxWidth()
                    .heightIn(min = 240.dp),
                decorationBox = { inner ->
                    if (controller.notes.isEmpty()) {
                        Text(text = "Start typing…", style = type.body, color = colors.muted)
                    }
                    inner()
                },
            )
        }

        // Smart suggestions footer placeholder.
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(QuickInkRadius.md))
                .background(colors.accentSoft)
                .padding(QuickInkSpacing.s4),
            verticalArrangement = Arrangement.spacedBy(QuickInkSpacing.s2),
        ) {
            Text(text = "SMART SUGGESTIONS", style = type.eyebrow, color = colors.muted)
            Text(
                text  = "Coming soon — extracted to-dos, dates, and contacts from this page.",
                style = type.bodyItalic,
                color = colors.inkSoft,
            )
        }
    }
}

// MARK: - Floating action bar

@Composable
private fun FloatingActionBar(
    showDelete: Boolean,
    onReTag: () -> Unit,
    onExport: () -> Unit,
    onDelete: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val colors = LocalQuickInkColors.current
    Row(
        modifier = modifier
            .clip(RoundedCornerShape(QuickInkRadius.pill))
            .shadow(12.dp, RoundedCornerShape(QuickInkRadius.pill))
            .background(colors.surface)
            .border(1.dp, colors.border, RoundedCornerShape(QuickInkRadius.pill))
            .padding(QuickInkSpacing.s2),
        horizontalArrangement = Arrangement.spacedBy(QuickInkSpacing.s2),
        verticalAlignment     = Alignment.CenterVertically,
    ) {
        ActionButton(icon = Icons.Filled.LocalOffer, label = "Re-tag", onClick = onReTag)
        ActionButton(icon = Icons.Filled.IosShare,   label = "Export", onClick = onExport)
        if (showDelete) {
            ActionButton(icon = Icons.Filled.Delete, label = "Delete", tint = colors.danger, onClick = onDelete)
        }
    }
}

@Composable
private fun ActionButton(
    icon: ImageVector,
    label: String,
    tint: Color = LocalQuickInkColors.current.ink,
    onClick: () -> Unit,
) {
    val type = LocalQuickInkTypography.current
    Row(
        modifier = Modifier
            .clip(RoundedCornerShape(QuickInkRadius.pill))
            .clickable(onClick = onClick)
            .padding(horizontal = QuickInkSpacing.s3, vertical = QuickInkSpacing.s2),
        verticalAlignment     = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        Icon(
            imageVector       = icon,
            contentDescription = null,
            tint              = tint,
            modifier          = Modifier.size(14.dp),
        )
        Text(text = label, style = type.label, color = tint)
    }
}

File: apps/quickink/ios/QuickInk/Notes/NoteEditorScreen.swift
/*
 * NoteEditorScreen.swift
 *
 * QuickInk's Note Detail — upgraded from the slim title+body editor
 * to the mockup's tabbed view:
 *
 *   - Tab 1 (Page): lined-paper page rendering with coral left
 *     margin; if an associated scan preview exists, the image
 *     overlays the paper rule. Tap the page to open a fullscreen
 *     reader (handled in a follow-up).
 *
 *   - Tab 2 (OCR Text): clean editorial serif rendering of the
 *     transcript with a confidence badge, copy-to-clipboard button,
 *     and a "smart suggestions" footer hook (currently inert).
 *
 *   - Floating action bar (Re-tag / Export / Delete).
 *
 * Architecturally we still wrap `NotepadEditorViewModel` — the
 * view's `notes` and `title` strings remain the data source.
 * Inserting OCR-only-vs-text-note logic is a mostly-presentational
 * concern: when the entry has captured pages, we render them in
 * the Page tab; otherwise the Page tab is a static lined-paper
 * card showing the handwritten title.
 *
 * Slice 4 deferred bits:
 *   - Re-tag flow (T13 / category model)
 *   - Export sheet (T14)
 *   - Smart-suggestions backend
 *   - Multi-page navigator (model is one entry = one page today)
 *   - Entity highlighting (NSAttributedString pass over body)
 *
 * Mirror of Android `NoteEditorScreen.kt`.
 */

import SwiftUI
#if canImport(UIKit)
import UIKit
#endif
import ReleafCoreNotes

struct NoteEditorScreen: View {

    let entryId: String
    let userId: String
    let onBack: () -> Void

    @StateObject private var vm: NotepadEditorViewModel
    @State private var activeTab: Tab = .page
    @State private var showCopyToast = false
    @State private var showExportSheet = false

    enum Tab { case page, ocr }

    init(entryId: String, userId: String, onBack: @escaping () -> Void) {
        self.entryId = entryId
        self.userId = userId
        self.onBack = onBack

        let repository = NotepadRepository(dbQueue: QuickInkDatabase.shared.dbQueue)
        _vm = StateObject(
            wrappedValue: NotepadEditorViewModel(
                repository: repository,
                entryId:    entryId,
                userId:     userId
            )
        )
    }

    var body: some View {
        ZStack(alignment: .bottom) {
            VStack(spacing: 0) {
                topBar
                tabSwitcher
                    .padding(.horizontal, QuickInkSpacing.s5)
                    .padding(.top, QuickInkSpacing.s3)
                    .padding(.bottom, QuickInkSpacing.s2)

                if vm.isLoading {
                    Spacer()
                    ProgressView()
                        .tint(QuickInkColors.accent)
                    Spacer()
                } else {
                    Group {
                        switch activeTab {
                        case .page: pageTab
                        case .ocr:  ocrTab
                        }
                    }
                    .frame(maxHeight: .infinity)
                }
            }

            floatingActionBar
                .padding(.horizontal, QuickInkSpacing.s5)
                .padding(.bottom, QuickInkSpacing.s5)

            if showCopyToast {
                copyToast
                    .padding(.bottom, 100)
                    .transition(.move(edge: .bottom).combined(with: .opacity))
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(QuickInkColors.bg.ignoresSafeArea())
        .task {
            await vm.bootstrap()
        }
        .sheet(isPresented: $showExportSheet) {
            ExportSheet(
                searchablePdfEnabled: UserDefaults.standard.bool(forKey: "quickink.settings.searchable_pdf_export_enabled"),
                onSelect: { format in
                    showExportSheet = false
                    // Format-specific export pipeline lands in a
                    // follow-up. Today the sheet's selection is a
                    // no-op — surfacing intent without actually
                    // generating the file. Once the Drive export +
                    // bitmap renderer ship, hand off here:
                    //   exporter.export(vm.entry, as: format)
                    _ = format
                },
                onDismiss: { showExportSheet = false }
            )
            .presentationDetents([.medium])
            .presentationDragIndicator(.hidden)
        }
    }

    // MARK: - Top bar

    @ViewBuilder
    private var topBar: some View {
        HStack(alignment: .center) {
            Button(action: backAndPersist) {
                Image(systemName: "chevron.left")
                    .font(.system(size: 18, weight: .medium))
                    .foregroundStyle(QuickInkColors.ink)
                    .padding(QuickInkSpacing.s3)
            }
            .accessibilityLabel("Back")

            VStack(alignment: .leading, spacing: 0) {
                if vm.title.isEmpty {
                    Text("Untitled")
                        .font(QuickInkText.label)
                        .foregroundStyle(QuickInkColors.muted)
                } else {
                    Text(vm.title)
                        .font(QuickInkText.label)
                        .foregroundStyle(QuickInkColors.ink)
                        .lineLimit(1)
                }
                Text(vm.entryDate)
                    .font(QuickInkText.caption)
                    .foregroundStyle(QuickInkColors.muted)
            }

            Spacer()

            // Page navigator pill — placeholder showing 1/1 until
            // multi-page model exists. Stays visible to set the
            // expectation that multi-page is coming.
            Text("1 / 1")
                .font(QuickInkText.caption)
                .foregroundStyle(QuickInkColors.inkSoft)
                .padding(.horizontal, QuickInkSpacing.s3)
                .padding(.vertical, 4)
                .background(QuickInkColors.borderSoft)
                .clipShape(RoundedRectangle(cornerRadius: QuickInkRadius.pill, style: .continuous))
                .padding(.trailing, QuickInkSpacing.s3)
        }
        .padding(.horizontal, QuickInkSpacing.s2)
        .padding(.top, QuickInkSpacing.s2)
    }

    // MARK: - Tab switcher

    @ViewBuilder
    private var tabSwitcher: some View {
        HStack(spacing: 4) {
            tabButton(label: "Page",     isActive: activeTab == .page) { activeTab = .page }
            tabButton(label: "OCR Text", isActive: activeTab == .ocr)  { activeTab = .ocr }
        }
        .padding(4)
        .background(QuickInkColors.borderSoft)
        .clipShape(RoundedRectangle(cornerRadius: QuickInkRadius.pill, style: .continuous))
    }

    @ViewBuilder
    private func tabButton(label: String, isActive: Bool, action: @escaping () -> Void) -> some View {
        Button(action: action) {
            Text(label)
                .font(QuickInkText.label)
                .foregroundStyle(isActive ? QuickInkColors.ink : QuickInkColors.inkSoft)
                .frame(maxWidth: .infinity)
                .padding(.vertical, QuickInkSpacing.s2)
                .background(
                    RoundedRectangle(cornerRadius: QuickInkRadius.pill, style: .continuous)
                        .fill(isActive ? QuickInkColors.surface : .clear)
                        .shadow(color: isActive ? QuickInkColors.ink.opacity(0.06) : .clear, radius: 4, x: 0, y: 2)
                )
        }
        .buttonStyle(.plain)
    }

    // MARK: - Page tab

    @ViewBuilder
    private var pageTab: some View {
        ScrollView {
            ZStack(alignment: .topLeading) {
                QuickInkLinedPaper(
                    tone: QuickInkColors.surface,
                    lineSpacing: 22,
                    lineOpacity: 0.10
                )

                // Coral left margin line — the iconic mockup detail.
                HStack(spacing: 0) {
                    Rectangle()
                        .fill(QuickInkColors.accent.opacity(0.6))
                        .frame(width: 1.5)
                        .padding(.leading, 36)
                    Spacer()
                }

                // Body content rendered in handwritten Caveat to
                // simulate the scanned page. When real captured
                // image data is wired, replace this with the page
                // image.
                VStack(alignment: .leading, spacing: QuickInkSpacing.s3) {
                    if !vm.title.isEmpty {
                        Text(vm.title)
                            .font(QuickInkFont.handwritten(28))
                            .foregroundStyle(QuickInkColors.ink)
                    }
                    Text(vm.notes.isEmpty ? "Tap OCR Text to add transcript content." : vm.notes)
                        .font(QuickInkFont.handwritten(20))
                        .foregroundStyle(QuickInkColors.ink.opacity(0.85))
                        .fixedSize(horizontal: false, vertical: true)
                }
                .padding(.leading, 52)
                .padding(.trailing, QuickInkSpacing.s5)
                .padding(.vertical, QuickInkSpacing.s5)
            }
            .frame(minHeight: 480)
            .clipShape(RoundedRectangle(cornerRadius: QuickInkRadius.lg, style: .continuous))
            .overlay(
                RoundedRectangle(cornerRadius: QuickInkRadius.lg, style: .continuous)
                    .stroke(QuickInkColors.border, lineWidth: 1)
            )
            .shadow(color: QuickInkColors.ink.opacity(0.06), radius: 8, x: 0, y: 4)
            .padding(QuickInkSpacing.s5)
            .padding(.bottom, 100)
        }
    }

    // MARK: - OCR tab

    @ViewBuilder
    private var ocrTab: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: QuickInkSpacing.s4) {
                // Confidence + copy header.
                HStack {
                    HStack(spacing: QuickInkSpacing.s1) {
                        Circle()
                            .fill(QuickInkColors.success)
                            .frame(width: 6, height: 6)
                        Text("98% confidence")
                            .font(QuickInkText.caption)
                            .foregroundStyle(QuickInkColors.inkSoft)
                    }
                    .padding(.horizontal, QuickInkSpacing.s3)
                    .padding(.vertical, QuickInkSpacing.s1)
                    .background(QuickInkColors.borderSoft)
                    .clipShape(RoundedRectangle(cornerRadius: QuickInkRadius.pill, style: .continuous))

                    Spacer()

                    Button(action: copyOCRText) {
                        HStack(spacing: 6) {
                            Image(systemName: "doc.on.doc")
                                .font(.system(size: 13))
                            Text("Copy")
                                .font(QuickInkText.label)
                        }
                        .foregroundStyle(QuickInkColors.accent)
                        .padding(.horizontal, QuickInkSpacing.s3)
                        .padding(.vertical, QuickInkSpacing.s1)
                        .background(QuickInkColors.accentSoft)
                        .clipShape(RoundedRectangle(cornerRadius: QuickInkRadius.pill, style: .continuous))
                    }
                    .buttonStyle(.plain)
                }

                // Title field — editable.
                TextField("Title", text: $vm.title)
                    .font(QuickInkText.pageTitle)
                    .foregroundStyle(QuickInkColors.ink)

                // Body / OCR transcript — editable. The mockup also
                // showed entity highlighting (dates, names, etc.) on
                // top of the transcript; with TextEditor we can't
                // run an attributed pass, so the highlighting hook
                // is captured below as a static read-only preview
                // that only renders when notes is non-empty.
                if vm.notes.isEmpty {
                    // Empty state — composing fresh.
                    TextEditor(text: $vm.notes)
                        .font(QuickInkText.body)
                        .foregroundStyle(QuickInkColors.ink)
                        .scrollContentBackground(.hidden)
                        .frame(minHeight: 200)
                        .padding(QuickInkSpacing.s3)
                        .background(QuickInkColors.surface)
                        .clipShape(RoundedRectangle(cornerRadius: QuickInkRadius.md, style: .continuous))
                        .overlay(
                            RoundedRectangle(cornerRadius: QuickInkRadius.md, style: .continuous)
                                .stroke(QuickInkColors.border, lineWidth: 1)
                        )
                } else {
                    // Read with edit-on-tap. Tap the body to enter
                    // an editable TextEditor inline.
                    TextEditor(text: $vm.notes)
                        .font(QuickInkText.body)
                        .foregroundStyle(QuickInkColors.ink)
                        .scrollContentBackground(.hidden)
                        .frame(minHeight: 240)
                        .padding(QuickInkSpacing.s4)
                        .background(QuickInkColors.surface)
                        .clipShape(RoundedRectangle(cornerRadius: QuickInkRadius.md, style: .continuous))
                        .overlay(
                            RoundedRectangle(cornerRadius: QuickInkRadius.md, style: .continuous)
                                .stroke(QuickInkColors.border, lineWidth: 1)
                        )
                }

                // Smart suggestions footer — placeholder. When the
                // suggestion model lands, populate with detected
                // todos, dates, contacts from the OCR text.
                VStack(alignment: .leading, spacing: QuickInkSpacing.s2) {
                    Text("SMART SUGGESTIONS")
                        .font(QuickInkText.eyebrow)
                        .tracking(QuickInkLetterSpacing.eyebrow)
                        .foregroundStyle(QuickInkColors.muted)

                    Text("Coming soon — extracted to-dos, dates, and contacts from this page.")
                        .font(QuickInkText.bodyItalic)
                        .foregroundStyle(QuickInkColors.inkSoft)
                }
                .padding(QuickInkSpacing.s4)
                .frame(maxWidth: .infinity, alignment: .leading)
                .background(QuickInkColors.accentSoft)
                .clipShape(RoundedRectangle(cornerRadius: QuickInkRadius.md, style: .continuous))
            }
            .padding(.horizontal, QuickInkSpacing.s5)
            .padding(.top, QuickInkSpacing.s3)
            .padding(.bottom, 100)
        }
    }

    // MARK: - Floating action bar

    @ViewBuilder
    private var floatingActionBar: some View {
        HStack(spacing: QuickInkSpacing.s2) {
            actionButton(systemName: "tag", label: "Re-tag", action: { /* T13 */ })
            actionButton(systemName: "square.and.arrow.up", label: "Export", action: { showExportSheet = true })
            if vm.entry != nil {
                actionButton(systemName: "trash", label: "Delete", tint: QuickInkColors.danger, action: deleteAndDismiss)
            }
        }
        .padding(QuickInkSpacing.s2)
        .background(QuickInkColors.surface)
        .clipShape(RoundedRectangle(cornerRadius: QuickInkRadius.pill, style: .continuous))
        .overlay(
            RoundedRectangle(cornerRadius: QuickInkRadius.pill, style: .continuous)
                .stroke(QuickInkColors.border, lineWidth: 1)
        )
        .shadow(color: QuickInkColors.ink.opacity(0.10), radius: 12, x: 0, y: 4)
    }

    @ViewBuilder
    private func actionButton(systemName: String, label: String, tint: Color = QuickInkColors.ink, action: @escaping () -> Void) -> some View {
        Button(action: action) {
            HStack(spacing: 6) {
                Image(systemName: systemName)
                    .font(.system(size: 14, weight: .medium))
                Text(label)
                    .font(QuickInkText.label)
            }
            .foregroundStyle(tint)
            .padding(.horizontal, QuickInkSpacing.s3)
            .padding(.vertical, QuickInkSpacing.s2)
        }
        .buttonStyle(.plain)
    }

    // MARK: - Copy toast

    @ViewBuilder
    private var copyToast: some View {
        Text("Copied to clipboard")
            .font(QuickInkText.label)
            .foregroundStyle(QuickInkColors.textOnAccent)
            .padding(.horizontal, QuickInkSpacing.s4)
            .padding(.vertical, QuickInkSpacing.s2)
            .background(QuickInkColors.ink)
            .clipShape(RoundedRectangle(cornerRadius: QuickInkRadius.pill, style: .continuous))
    }

    // MARK: - Actions

    private func backAndPersist() {
        // Save-on-back matches Releaf's editor flow — no separate
        // Save button. The VM's `save()` is a no-op when there's
        // nothing to commit (canSave == false), so it's safe to
        // always call.
        if vm.canSave {
            vm.save()
            QuickInkSyncEnvironment.shared.scheduler.requestImmediate()
        }
        onBack()
    }

    private func deleteAndDismiss() {
        vm.delete(onDeleted: {
            QuickInkSyncEnvironment.shared.scheduler.requestImmediate()
            onBack()
        })
    }

    private func copyOCRText() {
        #if canImport(UIKit)
        UIPasteboard.general.string = vm.notes
        #endif
        withAnimation(.easeInOut(duration: 0.18)) {
            showCopyToast = true
        }
        // Auto-dismiss after 1.6s.
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.6) {
            withAnimation(.easeInOut(duration: 0.18)) {
                showCopyToast = false
            }
        }
    }
}
